pkg/pillar/cmd/zedagent/handleconfig.go

```diff
@@ -150,6 +152,127 @@ var nilUUID uuid.UUID
+	}
+	changed, configSha, err := fileutils.CompareSha(
+		types.BootstrapConfFileName, types.BootstrapShaFileName)
+	if err != nil {
+		log.Errorf("CompareSha failed for bootstrap config: %s", err)
+		// We will not record SHA for applied bootstrap config
+		// and as a result load it again with the next boot.
+		// However, with config timestamp preventing accidental revert
+		// to an older configuration, this should not be a problem.
+		configSha = nil
+	} else if !changed {
+		// This bootstrap config was already applied.
+		return
+	}
+
+	// Load file content.
+	contents, err := ioutil.ReadFile(types.BootstrapConfFileName)
+	if err != nil {
+		log.Errorf("Failed to read bootstrap config: %v", err)
+		indicateInvalidBootstrapConfig(getconfigCtx)
+		return
+	}
+
+	// Mark bootstrap config as processed by storing SHA hash of its content
+	// under /persist/ingested/ directory.
+	// We do this even even if the unmarshalling (or anything else) below fails.
+	// This is to avoid repeated failing attempts to load invalid config on each boot.
+	defer func() {
+		if configSha != nil {
+			err := fileutils.SaveShaInFile(types.BootstrapShaFileName, configSha)
+			if err != nil {
+				log.Errorf("Failed to save SHA of bootstrap config: %v", err)
+			}
+		}
+	}()
+
+	// Unmarshal BootstrapConfig.
+	bootstrap := zconfig.BootstrapConfig{}
+	err = proto.Unmarshal(contents, &bootstrap)
+	if err != nil {
+		log.Errorf("Failed to unmarshal bootstrap config: %v", err)
+		indicateInvalidBootstrapConfig(getconfigCtx)
+		return
+	}
+
+	// Verify controller certificate chain.
+	sigCertBytes, err := zedcloud.VerifySigningCertChain(log, bootstrap.ControllerCerts)
+	if err != nil {
+		log.Errorf("Controller cert chain verification failed for bootstrap config: %v", err)
+		indicateInvalidBootstrapConfig(getconfigCtx)
+		return
+	}
+
+	// Verify payload signature
+	zedcloudCtx := zedcloud.NewContext(log, zedcloud.ContextOptions{})
+	if err = zedcloud.LoadServerSigningCert(&zedcloudCtx, sigCertBytes); err != nil {
+		log.Errorf("Failed to load signing server cert from bootstrap config: %v", err)
+		indicateInvalidBootstrapConfig(getconfigCtx)
+		return
+	}
+	_, err = zedcloud.VerifyAuthContainer(&zedcloudCtx, bootstrap.SignedConfig)
+	if err != nil {
+		log.Errorf("Signature verification failed for bootstrap config: %v", err)
+		indicateInvalidBootstrapConfig(getconfigCtx)
+		return
+	}
+
+	// Unmarshal EdgeDevConfig from the payload of AuthContainer.
+	devConfig := zconfig.EdgeDevConfig{}
+	payload := bootstrap.SignedConfig.GetProtectedPayload().GetPayload()
+	if payload == nil {
+		log.Error("Bootstrap config payload is nil")
+		indicateInvalidBootstrapConfig(getconfigCtx)
+		return
+	}
+	err = proto.Unmarshal(payload, &devConfig)
+	if err != nil {
+		log.Errorf("Failed to unmarshal bootstrap config payload: %v", err)
+		return
+	}
+
+	// Apply bootstrap config.
+	inhaleDeviceConfig(getconfigCtx, &devConfig, fromBootstrap)
```

Comment:
Does this have the effect of publishing ConfigItemValueMap even if there are no such items in the config?
Or is the parseConfigItem* conditional on there being some items to process?

In any case, need to document these key side effects, and I think the function should have a different name like "maybeProcessBootstrapConfig()

Reply:
This is documented here: https://github.com/milan-zededa/eve/blob/bootstrap-config/pkg/pillar/cmd/zedagent/parseconfig.go#L2281-L2304
Basically zedagent merges default values with whatever is set in the received config. It then compares the new value with the previous config and publishes if a change was detected. I have made small improvement in the PR to ensure that zedagent tries to publish also when the previous config was not published successfully due to some (transient) error.
Also `loadBootstrapConfig` was renamed and suggested and comment was added to describe why it is conditional.